File: 1001-1500/1100_numKLenSubstrNoRepeats.py
class Solution:
    def numKLenSubstrNoRepeats(self, s: str, k: int) -> int:
        frequency = {}
        count = 0
        for i, char in enumerate(s):
            if char not in frequency:
                frequency[char] = 1
            else:
                frequency[char] += 1
            if i >= k:
                frequency[s[i - k]] -= 1
                if not frequency[s[i - k]]:
                    del frequency[s[i - k]]
            if len(frequency) == k:
                count += 1
        return count

File: 1001-1500/1100_numKLenSubstrNoRepeats.java
class Solution {
    public int numKLenSubstrNoRepeats(String s, int k) {
        if (s.length() < k) return 0;
        HashMap<Character, Integer> window = new HashMap<>();
        int count = 0;
        for (int i=0; i<k; i++) window.put(s.charAt(i), window.getOrDefault(s.charAt(i), 0) + 1);
        if (window.size() == k) count++;
        for (int i=k; i<s.length(); i++) {
            window.put(s.charAt(i-k), window.get(s.charAt(i-k)) - 1);
            if (window.get(s.charAt(i-k)) == 0) window.remove(s.charAt(i-k));
            window.put(s.charAt(i), window.getOrDefault(s.charAt(i), 0) + 1);
            if (window.size() == k) count++;
        }
        return count;
    }
}
